File: kotlin-node/karakum/src/jsMain/resources/plugins/convertOverriddenFsOptionPropertySignature.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default function (node, context, render) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("fs.d.ts") && !sourceFileName.endsWith("fs/promises.d.ts")) return null

    if (
        ts.isPropertySignature(node)
        && ts.isTypeLiteralNode(node.parent)
        && (
            node.name.text === "bigint"
            || node.name.text === "throwIfNoEntry"
            || (
                node.name.text === "recursive"

                && ts.isIntersectionTypeNode(node.parent.parent)
                && node.parent.parent.types[0]
                && ts.isTypeReferenceNode(node.parent.parent.types[0])
                && ts.isIdentifier(node.parent.parent.types[0].typeName)
                && node.parent.parent.types[0].typeName.text === "MakeDirectoryOptions"
            )
            || (
                node.name.text === "encoding"

                && ts.isIntersectionTypeNode(node.parent.parent)
                && node.parent.parent.types[0]
                && ts.isTypeReferenceNode(node.parent.parent.types[0])
                && ts.isIdentifier(node.parent.parent.types[0].typeName)
                && node.parent.parent.types[0].typeName.text === "WatchOptions"
            )
        )
    ) {
        const readonly = node.modifiers?.find(modifier => modifier.kind === ts.SyntaxKind.ReadonlyKeyword)

        const modifier = readonly
            ? "val "
            : "var "

        const name = render(node.name)

        const isOptional = Boolean(node.questionToken)

        const type = karakum.renderNullable(node.type, isOptional, context, render)

        const hasAlias = !isOptional

        if (hasAlias) {
            const alias = node.name.text === "bigint"
                ?  "requiredBigInt"
                : `required${karakum.capitalize(name)}`
            const jsName = `@JsName("${name}")`
            const aliasSignature = `${modifier}${alias}: ${type}`

            return [
                jsName,
                aliasSignature,
            ].join("\n")
        }

        return `override ${modifier}${name}: ${type}`
    }
    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertOverriddenFsOptionPropertySignature.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.renderNullable
import io.github.sgrishchenko.karakum.util.capitalize
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.SyntaxKind
import typescript.asArray
import typescript.isIdentifier
import typescript.isIntersectionTypeNode
import typescript.isPropertySignature
import typescript.isTypeLiteralNode
import typescript.isTypeReferenceNode

val convertOverriddenFsOptionPropertySignature = createPlugin { node, context, render ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(
            sourceFileName.endsWith("fs.d.ts")
                    || sourceFileName.endsWith("fs/promises.d.ts")
        )

        ensure(isPropertySignature(node))

        val propertyName = node.name
        ensure(isIdentifier(propertyName))

        val typeLiteral = ensureNotNull(node.getParentOrNull())
        ensure(isTypeLiteralNode(typeLiteral))

        ensureNotNull(
            nullable {
                ensure(
                    propertyName.text == "bigint"
                            || propertyName.text == "throwIfNoEntry"
                )
            } ?: nullable {
                ensure(propertyName.text == "recursive")

                val intersection = ensureNotNull(typeLiteral.getParentOrNull())
                ensure(isIntersectionTypeNode(intersection))

                val firstType = ensureNotNull(intersection.types.asArray().firstOrNull())
                ensure(isTypeReferenceNode(firstType))

                val typeName = firstType.typeName
                ensure(isIdentifier(typeName))
                ensure(typeName.text == "MakeDirectoryOptions")
            } ?: nullable {
                ensure(propertyName.text == "encoding")

                val intersection = ensureNotNull(typeLiteral.getParentOrNull())
                ensure(isIntersectionTypeNode(intersection))

                val firstType = ensureNotNull(intersection.types.asArray().firstOrNull())
                ensure(isTypeReferenceNode(firstType))

                val typeName = firstType.typeName
                ensure(isIdentifier(typeName))
                ensure(typeName.text == "WatchOptions")
            }
        )

        val readonly = node.modifiers?.asArray()?.find { it.kind == SyntaxKind.ReadonlyKeyword }

        val modifier = if (readonly != null) "val " else "var "

        val name = render(node.name)

        val isOptional = node.questionToken != null

        val type = renderNullable(node.type, isOptional, context, render)

        val hasAlias = !isOptional

        if (hasAlias) {
            val alias = if (propertyName.text == "bigint") {
                "requiredBigInt"
            } else {
                "required${capitalize(name)}"
            }
            val jsName = "@JsName(\"$name\")"
            val aliasSignature = "${modifier}${alias}: $type"

            arrayOf(
                jsName,
                aliasSignature,
            ).joinToString("\n")
        } else {
            "override ${modifier}${name}: $type"
        }
    }
}
